pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -86,90 +101,370 @@ func (c *ClusterMemberController) sync(ctx context.Context, syncCtx factory.Sync
 func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder events.Recorder) error {
 	unhealthyMembers, err := c.etcdClient.UnhealthyMembers(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get list of unhealthy members: %v", err)
+		return err
 	}
 	if len(unhealthyMembers) > 0 {
-		klog.V(4).Infof("unhealthy members: %v", spew.Sdump(unhealthyMembers))
+		klog.V(4).Infof("reconcile aborted found unhealthy etcd members: %v", spew.Sdump(unhealthyMembers))
 		return fmt.Errorf("unhealthy members found during reconciling members")
 	}
 
-	// etcd is healthy, decide if we need to scale
-	podToAdd, err := c.getEtcdPodToAddToMembership(ctx)
-	switch {
-	case err != nil:
-		return fmt.Errorf("could not get etcd pod: %w", err)
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
-	default:
-		recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
+	// Set desired cluster size
+	// TODO(haseeb): Always update this to react to control plane size change in cluster config?
+	if c.replicaCount == 0 {
+		replicaCount, err := ceohelpers.GetMastersReplicaCount(ctx, c.kubeClient)
+		if err != nil {
+			return fmt.Errorf("failed to get control-plane replica count: %w", err)
+		}
+		c.replicaCount = replicaCount
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Unblock Machines pending deletion by removing their learner members and PreDrain hooks.
+	// Add PreDrain hooks to new machines
+	err = c.reconcileMachineHooks(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile machine hooks: %v", err)
+	}
+
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		return fmt.Errorf("learner promotion failed: %w", err)
+	}
+
+	// etcd is healthy, decide if we need to scale.
+	peerURL, err := c.getEtcdPeerURLToScaleUp(ctx)
 	if err != nil {
 		return fmt.Errorf("could not get etcd peer host :%w", err)
 	}
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s:2380", etcdHost))
+	// Scale up if next peer found
+	if peerURL != "" {
+		recorder.Eventf("ScaleUpCluster", "scale up attempt peerURl: %s", peerURL)
+		err = c.etcdClient.MemberAdd(ctx, peerURL)
+		if err != nil {
+			recorder.Warningf("ScaleUpCluster", "scale up failed peerURl: %s :%v", peerURL, err)
+			return err
+		}
+		recorder.Eventf("ScaleUpCluster", "scale up etcd learner success peerURl: %s", peerURL)
+	}
+
+	memberName, err := c.getEtcdMemberToScaleDown(ctx)
 	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+		return fmt.Errorf("could not get etcd peer host :%w", err)
+	}
+	// Scale down if next peer found
+	if memberName != "" {
```

Comment:
Looks to me like `getEtcdMemberToScaleDown` has been adjusted to never return a name, I'm guessing this if statement is going to go away at some point?